src/main/scala/util/Property.scala

```diff
@@ -0,0 +1,45 @@
+// See LICENSE.SiFive for license details.
+
+package freechips.rocketchip.util.property
+
+import Chisel._
+import chisel3.internal.sourceinfo.{SourceInfo, SourceLine}
+import freechips.rocketchip.config.{Field, Parameters}
+
+case object PropertyLibrary extends Field[BasePropertyLibrary[BasePropertyParameters]]
+
+
+sealed abstract class PropertyType(name: String) {
+  override def toString: String = name
+}
+
+object PropertyType {
+  object Assert extends PropertyType("Assert")
+  object Assume extends PropertyType("Assume")
+  object Cover extends PropertyType("Cover")
+}
+
+class BasePropertyParameters(val p_type: PropertyType, val cond: Bool, val label: String, val message: String) {}
+
+class BasePropertyLibrary[PropParam <: BasePropertyParameters] {
+  def generateCover(prop_param: PropParam)(implicit sourceInfo: SourceInfo): Unit = {
+    // Currently empty for user to extend their own cover methods
+    Unit
+  }
+}
```

Comment:
i would turn this into an abstract class:
```
abstract class BasePropertyLibrary[PP <: BasePropertyParameters] {
  def generateCover(pp: PP)(implicit sourceInfo: SourceInfo): Unit
}
```
and then maybe add a default concrete class that does something more specific:
```
class PrintCoverPropertyLibrary extends BasePropertyLibrary[CoverPropertyParameters] {
   def generateCover(pp: CoverPropertyParameters)(implicit sourceInfo: SourceInfo) { printf(p.message) }
}
```

Reply:
If you're returning Unit stylistically omit the `...): Unit = {...}` and just write `...) {...} `

Reply:
Should `generateCover` not admit instances of `BasePropertyParameters` where `p_type != Cover`?